File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/shop/ShopDsl.java
package com.optivem.shop.dsl.core.usecase.shop;

import com.optivem.shop.dsl.driver.port.shop.ShopDriver;
import com.optivem.shop.dsl.core.usecase.shop.usecases.BrowseCoupons;
import com.optivem.shop.dsl.core.usecase.shop.usecases.CancelOrder;
import com.optivem.shop.dsl.core.usecase.shop.usecases.DeliverOrder;
import com.optivem.shop.dsl.core.usecase.shop.usecases.GoToShop;
import com.optivem.shop.dsl.core.usecase.shop.usecases.PlaceOrder;
import com.optivem.shop.dsl.core.usecase.shop.usecases.PublishCoupon;
import com.optivem.shop.dsl.core.usecase.shop.usecases.ViewOrder;
import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.core.shared.UseCaseContext;

import java.io.Closeable;

public class ShopDsl implements Closeable {
    private final ShopDriver driver;
    private final UseCaseContext context;

    public ShopDsl(ShopDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToShop goToShop() {
        return new GoToShop(driver, context);
    }

    public PlaceOrder placeOrder() {
        return new PlaceOrder(driver, context);
    }

    public CancelOrder cancelOrder() {
        return new CancelOrder(driver, context);
    }

    public DeliverOrder deliverOrder() {
        return new DeliverOrder(driver, context);
    }

    public ViewOrder viewOrder() {
        return new ViewOrder(driver, context);
    }

    public PublishCoupon publishCoupon() {
        return new PublishCoupon(driver, context);
    }

    public BrowseCoupons browseCoupons() {
        return new BrowseCoupons(driver, context);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/shop/ShopDsl.ts
import type { ShopDriver } from '../../../../driver/port/shop/shop-driver.js';
import type { UseCaseContext } from '../../use-case-context.js';
import { PlaceOrder } from './usecases/PlaceOrder.js';
import { ViewOrder } from './usecases/ViewOrder.js';
import { CancelOrder } from './usecases/CancelOrder.js';
import { DeliverOrder } from './usecases/DeliverOrder.js';
import { BrowseCoupons } from './usecases/BrowseCoupons.js';
import { PublishCoupon } from './usecases/PublishCoupon.js';
import { GoToShop } from './usecases/GoToShop.js';

export class ShopDsl {
  constructor(
    private readonly driver: ShopDriver,
    private readonly context: UseCaseContext,
  ) {}

  goToShop(): GoToShop {
    return new GoToShop(this.driver, this.context);
  }

  placeOrder(): PlaceOrder {
    return new PlaceOrder(this.driver, this.context);
  }

  viewOrder(): ViewOrder {
    return new ViewOrder(this.driver, this.context);
  }

  cancelOrder(): CancelOrder {
    return new CancelOrder(this.driver, this.context);
  }

  deliverOrder(): DeliverOrder {
    return new DeliverOrder(this.driver, this.context);
  }

  browseCoupons(): BrowseCoupons {
    return new BrowseCoupons(this.driver, this.context);
  }

  publishCoupon(): PublishCoupon {
    return new PublishCoupon(this.driver, this.context);
  }

  async close(): Promise<void> {
    await this.driver.close();
  }
}
